[RFC, RFT] mvebu: add 6.6 testing kernel

Borromini opened the pull request:
This adds 6.6 support for mvebu.

 Compile-tested:
 * mvebu/cortexa9: WRT1200AC
 * mvebu/cortexa72:  RB5009UG+S+IN

Run-tested:
* mvebu/cortexa9: WRT1200AC (myself), WRT3200ACM (@anomeome), Turris Omnia (@drujd)
* mvebu/cortexa53: GL-MV1000 (@mrkiko)
* mvebu/cortexa72: RB5009UG+S+IN (myself)

Borromini commented:
CI is failing on the DTB move, but that's a given since 6.1 has the DTSes still in the main dts/ dir, and CI is compiling 6.1. Not sure how to have the CI compile 6.6 short of switching kernel version already.

namiltd commented:
Take a look at this: https://github.com/openwrt/openwrt/pull/14713
Use kernel_bump.sh to avoid losing file history.

anomeome commented:
DNWFM with [configdiff](https://termbin.com/cnz0), fail is:

```
arm-openwrt-linux-muslgnueabi-nm -n /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/linux-6.6.20/vmlinux.o | awk '/^[0-9a-f]+ [rR] __ksymtab_/ {print substr($3,11)}' > /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/kernel_symtab.txt
grep -Ff /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/mod_symtab.txt /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/kernel_symtab.txt > /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/sym_include.txt
make[4]: *** [Makefile:24: /home/kc/wrtpac/source/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/linux-mvebu_cortexa9/symtab.h] Error 1

```
_symtab.h_ extent, but zero length...

mrkiko commented:
Tested and running successfully on GL-MV1000 (initramfs) altough with some emitted warnings related to DTS
See http://sprunge.us/htgN2k
for dmesg and some other sparse informations. Feel free to contact me on IRC.
The build prompted for some AEAD options - relative to AEAD split I guess.

Borromini commented:
@anomeome I saw that happen too, a `make dirclean` made it go away though. Haven't been able to reproduce it since.

Borromini commented:
> The build prompted for some AEAD options - relative to AEAD split I guess.

Can you share the symbols you had to set? So I can add them to the cortexa53 config.

anomeome commented:
@Borromini , Yes, that is where I started but no success.

mrkiko commented:
CRYPTO_SM4_ARM64_CE_CCM
CONFIG_CRYPTO_SM4_ARM64_CE_GCM

Borromini commented:
> CRYPTO_SM4_ARM64_CE_CCM CONFIG_CRYPTO_SM4_ARM64_CE_GCM

I will add them to the A53 config, thanks.

@robimarko Would you mind reviewing this? I'm curious about your take on the DTS vendor subdir issue as well. Thanks.

BKPepe commented on target/linux/mvebu/patches-6.6/700-mvneta-tx-queue-workaround.patch:
```diff
@@ -0,0 +1,43 @@
+From: Felix Fietkau <nbd@nbd.name>
+Subject: mvneta: tx queue workaround
```
Do we need still this hacky patch for kernel 6.6?

anomeome replied:
reference [5411](https://github.com/openwrt/openwrt/issues/5411)

BKPepe replied:
That reference is in the patch description, but it does not answer my question. :-)

stklcode replied:
Let's get some empirical data. I repeated my tests from 6.1 bump (https://github.com/openwrt/openwrt/pull/12938#issuecomment-1599133250).

**Test device:**
Linksys WRT1900ACS (mvebu/cortexa9)
OpenWrt SNAPSHOT, r0+25621-231d84e7a8 (f7c732bf9ed9bd09646a528681256fb7b0d80bfc with this PR on top)
Kernel 6.6.22
netperf 2.7.0-r3

**Test client:**
x86_64 Fedora 39 workstation
intel I219-V GBit Ethernet directly connected
netperf 2.7.1

**Test commands**
```
netperf -H 172.16.101.1 -l 60 -t TCP_MAERTS &
netperf -H 172.16.101.1 -l 60 -t TCP_MAERTS &
netperf -H 172.16.101.1 -l 60 -t TCP_MAERTS &
netperf -H 172.16.101.1 -l 60 -t TCP_MAERTS &
```

**Results with patch 700**

```
Recv   Send    Send                          
Socket Socket  Message  Elapsed              
Size   Size    Size     Time     Throughput  
bytes  bytes   bytes    secs.    10^6bits/sec
131072  16384  16384    60.00     234.15
131072  16384  16384    60.00     234.14
131072  16384  16384    60.00     234.14
131072  16384  16384    60.00     234.14
                                  936,57   (total)
```


**Results without patch 700**

```
Recv   Send    Send                          
Socket Socket  Message  Elapsed              
Size   Size    Size     Time     Throughput  
bytes  bytes   bytes    secs.    10^6bits/sec
131072  16384  16384    60.00     888.79
131072  16384  16384    60.01      15.95
131072  16384  16384    60.01      15.95
131072  16384  16384    60.01      15.92
                                  936,61   (total)
```

(both results repeatable with minor deviations)

**Conclusion**

Sadly the answer is "Yes", patch seems to be still necessary.

BKPepe replied:
Hmm, it would be interesting to see if it is even broken by using vanilla kernel, but I think that upstream Linux kernel developers would notice that.  Does anyone try to reach them? Isn't this caused for us by another hacky patch in this repository?

anomeome replied:
Is software flow offload is working with 6.6.x?

drujd replied:
> Is software flow offload is working with 6.6.x?

From what I've gathered, it does not on devices that use the switch for both LAN and WAN, but it does for e.g. Turris Omnia which has a separate WAN connection. It is working on my Turris Omnia - with the usual quirks.

robimarko commented:
@Borromini For the DTS path change, see how mediatek solved it
https://github.com/openwrt/openwrt/commit/7d1145ecb97b9c7f852714c7938771d80b2df4e8

drujd commented:
So far seems to work well on Turris Omnia (cortexa9/armv7).

Borromini committed:
mvebu: cortexa9: set DTS dir for 6.6

With 6.6, all DTSes were moved to their vendor subdirectories. ARM64
DTSes already used this scheme, but 32 bit Cortex A9 did not, prior
to 6.6. Introduce a kernel version check to keep backward compatibility
with 6.1.

Suggested-by: Robert Marko <robimarko@gmail.com>
Signed-off-by: Stijn Segers <foss@volatilesystems.org>

Borromini commented:
> @Borromini For the DTS path change, see how mediatek solved it [7d1145e](https://github.com/openwrt/openwrt/commit/7d1145ecb97b9c7f852714c7938771d80b2df4e8)

@robimarko Thanks for the hint. I have added the if clause, which works for the upstream DTSes, but that would make the buildroot think our OpenWrt specific DTSes were under `$(DTS_DIR)/marvell` as well, while they're statically at `files/arch/arm/boot/dts/`, no matter which kernel, 6.1 or 6.6. So I moved our downstream Cortex A9 DTSes into a `dts/` subdir just like the mediatek target. I hope that's acceptable. Tested with both kernels, works like it should.

If that's okay, I'll push those changes to the PR. Haven't had the time to look into the kmod-crypto stuff yet.

@anomeome While testing I ran into the symtab.h issue again as well, it looks like some config option is causing it. Using default buildroot options (ie wipe config, just select a single or multiple device(s) and don't touch anything else besides ticking the testing kernel), it compiles just fine. Hopefully that can help troubleshoot.

robimarko commented:
@Borromini No, that would break 6.1 kernel, just make `files-6.6` and put the DTS-es in the new path there, while adding `files-6.1` with the old path

Borromini commented:
I compile-tested both 6.1 and 6.6 and it won't. I can link the patch this evening for testing. A DTS dir seems cleaner to me than files-6.1 and files-6.6, IMHO?

robimarko commented:
dts was traditionally used when there was no upstream dts directory for that vendor/platform, files is more standard

Borromini commented:
Got it. Will do the files-6.1/6.6 split then.

Borromini commented:
@anomeome As per https://github.com/openwrt/openwrt/issues/14860, can you try compiling without `CONFIG_STRIP_KERNEL_EXPORTS`?

anomeome commented:
Yep. Compiled all the _wrtpac_ targets, flashed and running on a _rango_.; I saw a post on the forum where someone said they had to go back to OEM to move forward with their 6.6.x build, I did not see this issue.

Borromini committed:
mvebu: move files to files-6.1

DTS paths for 32 bit ARM devices changed with 6.6, move files/ to
files-6.1 to prep for kernel 6.6 introduction.

Signed-off-by: Stijn Segers <foss@volatilesystems.org>

mrkiko commented:
PR #14952 by @robimarko adds missing symbols.

Borromini commented:
Thanks, tree did show indeed lots of other ARM targets already manipulated those symbols in their config files.

PR is refreshed, using files-6.{1,6} as requested.

dannil commented:
Run-tested on WRT1900ACS, it's a spare unit so can test any additional changes if needed.

anomeome commented:
@Borromini , just a data point that _mamba_ is again close to limit (~72K), I would guess it will last until next major kernel push.

Borromini committed:
mvebu: 6.6: copy files, patches & configs from 6.1

Copy all mvebu 6.1 specific files, patches and configs to 6.1.

Signed-off-by: Stijn Segers <foss@volatilesystems.org>

Borromini committed:
mvebu: 6.6: adjust 32 bit ARM DTS path

As of 6.6, all upstream DTSes are moved to their respective vendor subdir.
OpenWrt already followed this practice for ARM64, but not yet for 32 bit
ARM (Armada 37x/38x).

Signed-off-by: Stijn Segers <foss@volatilesystems.org>

Borromini committed:
mvebu: 6.6: remove upstreamed patches

000-cpufreq-armada-8k-add-ap807-support.patch was upstreamed.

Signed-off-by: Stijn Segers <foss@volatilesystems.org>

Borromini committed:
mvebu: add 6.6 testing kernel

Add 6.6 testing kernel for mvebu target.

Signed-off-by: Stijn Segers <foss@volatilesystems.org>

Borromini commented:
I've come to realise all changes thrown together in [one single commit](https://github.com/openwrt/openwrt/pull/14912/commits/35598b3486d4abf922562ec07de204e9e41ca636) is a bit annoying if troubleshooting is needed, any objections if I refresh this PR with the changes split out?

robimarko commented:
Yeah, I was about to suggest the same

Borromini commented:
Done. Retested on my devices (WRT1200AC, RB5009UG+S+IN), all good.

mrkiko commented:
Regarding the dmesg output I sent, do you think we have to worry about the error messages related to phylink, if I am correct?

Borromini commented:
> Regarding the dmesg output I sent, do you think we have to worry about the error messages related to phylink, if I am correct?

Can you be more specific? I don't see any issues at first sight there.

mrkiko commented:
On Mon, Mar 25, 2024 at 01:51:13AM -0700, Borromini wrote:
> > Regarding the dmesg output I sent, do you think we have to worry about the error messages related to phylink, if I am correct?
> 
> Can you be more specific? I don't see any issues at first sight there.

Here is a selection of messages from my side, some of these are clearly not related to the PR or not problematic, just decided to leave them

[    0.000000] OF: fdt: Reserved memory: failed to reserve memory for node ***@***.***': base 0x0000000004400000, size 16 MiB
[    0.556411] debugfs: Directory 'd0060900.xor' with parent 'dmaengine' already present!
[    0.624646] OF: Bad cell count for ***@***.******@***.******@***.***/partitions
[    0.657280] mv88e6085 d0032004.mdio-mii:01: switch 0x3400 detected: Marvell 88E6141, revision 0
[    0.679560] mvneta d0030000.ethernet eth0: Using device tree mac address 94:83:c4:03:ae:ad
[    0.896580] mv88e6085 d0032004.mdio-mii:01: switch 0x3400 detected: Marvell 88E6141, revision 0
[    0.927088] hwmon hwmon0: temp1_input not attached to any thermal zone
[    0.945722] hwmon hwmon1: temp1_input not attached to any thermal zone
[    0.964033] hwmon hwmon2: temp1_input not attached to any thermal zone
[    1.269140] mv88e6085 d0032004.mdio-mii:01: OF node ***@***.******@***.******@***.******@***.*** of CPU port 0 lacks the required "phy-mode" property
[    1.284064] mv88e6085 d0032004.mdio-mii:01: OF node ***@***.******@***.******@***.******@***.*** of CPU port 0 lacks the required "phy-handle", "fixed-link" or "managed" properties
[    1.301657] mv88e6085 d0032004.mdio-mii:01: Skipping phylink registration for CPU port 0
[    1.389087] mv88e6085 d0032004.mdio-mii:01 wan (uninitialized): PHY ***@***.******@***.******@***.***!mdio:11] driver [Marvell 88E6341 Family] (irq=54)
[    1.488768] mv88e6085 d0032004.mdio-mii:01 lan0 (uninitialized): PHY ***@***.******@***.******@***.***!mdio:12] driver [Marvell 88E6341 Family] (irq=55)
[    1.588747] mv88e6085 d0032004.mdio-mii:01 lan1 (uninitialized): PHY ***@***.******@***.******@***.***!mdio:13] driver [Marvell 88E6341 Family] (irq=56)
[    1.699922] turris-mox-rwtm firmware:armada-3700-rwtm: Cannot read board information: -5
[    1.708039] armada-37xx-rwtm-mailbox d00b0000.mailbox: Secure processor not ready
[    1.715691] turris-mox-rwtm firmware:armada-3700-rwtm: Firmware does not support the GET_RANDOM command
[    1.725146] turris-mox-rwtm: probe of firmware:armada-3700-rwtm failed with error -5
[    2.103434] gpio_button_hotplug: loading out-of-tree module taints kernel.

> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/14912#issuecomment-2017492145
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

robimarko commented:
Those warnings are quite clear, that device DTS was not done according to the bindings and they just added validation

mrkiko commented:
On Mon, Mar 25, 2024 at 02:27:00AM -0700, Robert Marko wrote:
> Those warnings are quite clear, that device DTS was not done according to the bindings and they just added validation

Got it, thanks!

> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/14912#issuecomment-2017560104
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>

robimarko commented:
Please no commits without the accompanying description

Borromini commented:
Sorry, should be okay now.

stklcode commented:
Looks like there are still a few commits without description, e.g.

```
mvebu: 6.6: update Linksys status LED patch

Signed-off-by: Stijn Segers <foss@volatilesystems.org>
```

If those `update ... patch` commits are just manual refreshs, I personally prefer to squash all the refreshs together and add just one message instead of adding no additional information like "update for 6.6" to every single one

```
mvebu: 6.6: refresh patches

manually refreshed:
* 309-linksys-status-led.patch
* 310-linksys-use-eth0-as-cpu-port.patch
* 320-arm-dts-armada-370-synology-ds213j-mtd-parts.patch
* 701-mvpp2-read-mac-address-from-nvmem.patch
* 902-drivers-mfd-Add-a-driver-for-IEI-WT61P803-PUZZLE-MCU.patch

all other patches automatically refreshed

Signed-off-by: ...
```

Borromini committed:
mvebu: 6.6: refresh patches

Manually refreshed:
 * 309-linksys-status-led.patch
 * 310-linksys-use-eth0-as-cpu-port.patch
 * 320-arm-dts-armada-370-synology-ds213j-mtd-parts.patch
 * 701-mvpp2-read-mac-address-from-nvmem.patch
 * 902-drivers-mfd-Add-a-driver-for-IEI-WT61P803-PUZZLE-MCU.patch

All other patches automatically refreshed.

Signed-off-by: Stijn Segers <foss@volatilesystems.org>

Borromini commented:
Thanks, that makes more sense. Done.

robimarko commented:
You still have a commit without desciption, CI is falling as well on it due to formal issues

Borromini commented:
Well that's the sleep deprivation I guess :|. Should be good now.

mrkiko commented:
Tested PR in it's current state in the GL-MV1000 after rebasing in main (GCC 13). Works fine.

stklcode commented:
Build and run tested again on WRT1900ACS and Turris Omnia (both mvebu/cortexa9) with some custom configuraiton, dual MT7915e WiFi, EP06 LTE modem and WireGuard. No regressions so far.

Borromini commented:
@robimarko I think the comments have been addressed, can this be merged?

namiltd commented:
Some files (e.g. target/linux/mvebu/patches-6.6/300-mvebu-Mangle-bootloader-s-kernel-arguments.patch) require updating to kernel 6.6.23.
Please refresh patches.
